### ISSUE TITLE
Comparison to text-generation-webui

### ISSUE THREAD

**Issue by paplorinc** [zetavg/LLaMA-LoRA-Tuner#18]
I love what you've created, was wondering how it differs from the efforts at https://github.com/camenduru/text-generation-webui, which also seems to be able to tune lora models https://github.com/oobabooga/text-generation-webui/blob/main/docs/Training-LoRAs.md

**Comment by nRuaif** [comment]
For me, LoRA training in webui always results in OOM when saving LoRA

**Comment by zetavg** [comment]
For short, I think this project will be more focus on the train-evaluation loop, as my own use-case.

To my latest knowledge of `text-generation-webui`, this project has several notable drawbacks:

* Less generation parameters can be set on the UI.
* No support for presets for generation parameters.
* Lack of generation controls: Continue, Count tokens, etc.
* Can’t display results as Markdown/HTML
* No “Chat UI” and “Notebook mode”.
* Model loading settings can’t be adjusted on the UI, as well as loading multiple LoRA models isn’t supported.
* Limited UI controls for fine-tuning. However, a JSON input in the UI allows setting any supported trainer or LoRA configuration hyperparameters from Hugging Face Transformers.
* Uncertain local hardware support as I’ve only used cloud GPUs so far.
* Things might not have been done right since I’m new to this field.

Although it has some minor advantages as well:

* The training status is “synchronized across different devices” so that I can start the training on my computer, then check the progress on my phone, and also abort it to adjust the parameters and restart the training on a phone if necessary. (Recently added)
* Supports logging to W&B.
* Shows the % of trainable params and a training loss chart on the UI. (Recently added)
* Support *Flagging* inference outputs for later examination. Also shows/save the output tensors of the model.
* Mobile friendly, maybe.
* Support “hot-swapping” models (moves unused model from VRAM to RAM when switching between models). This is useful for comparing different models on machines with large RAMs (40GB+) since switching between models can be a matter of seconds instead of tens-of-seconds or even minutes.
* Loads models directly from Hugging Face, which leverages the same cache uses by Hugging Face Transformers, which is convent to play with the models directly in Python without maintaining two copies of the same model.
* Has templates to run on Google Colab and Lambda Cloud, GCP via SkyPilot.

Long version and thoughts:

This is a fun project created to aside my study of language models, which later became a tool that I can use to drag others to get started playing with language models so that I’ll have someone to discuss problems with. lol

As one of my goals is to find a cost-efficient way to fine-tune existing 6~13B LLMs which are trained in English, to support communications in `zh-tw`, and hopefully can carry their knowledge learned from English to `zh-tw`, I think I’ll add the ability to train tokenizers and support CLI training so that long training tasks can be done with managed spot instance. I might do a major re-work on the fine-tuning functionality as well to support training on dialogue data (while user inputs masked out) and a more flexible “prompter”/“formatter” solution (Has been reading the source code of https://github.com/lm-sys/FastChat recently, or maybe I’ll just use it). And besides this, there’re some features I wish to have as well, listed at: https://github.com/zetavg/LLaMA-LoRA-Tuner/discussions/14

Even though I’ll be happy to do so, given that I have limited time on my studies and work on these projects, I feel that the chances are small that I’ll make much contributions to related projects in the near future. But I’ll be happy to see if someone found something useful here and take/port them to other open-source projects, or vice-versa.

**Comment by paplorinc** [comment]
What a response <3
Keep up the good work!
